File: backend/src/test/java/lan/dk/podcastserver/business/ItemBusinessTest.java
package lan.dk.podcastserver.business;

import com.github.davinkevin.podcastserver.service.MimeTypeService;
import com.querydsl.core.types.Predicate;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.entity.Status;
import lan.dk.podcastserver.entity.Tag;
import lan.dk.podcastserver.manager.ItemDownloadManager;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.springframework.data.domain.*;
import org.springframework.util.FileSystemUtils;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.IntStream;

import static io.vavr.API.Set;
import static lan.dk.podcastserver.assertion.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

/**
 * Created by kevin on 02/08/15 for Podcast Server
 */
@RunWith(MockitoJUnitRunner.class)
public class ItemBusinessTest {

    private String ROOT_FOLDER = "/tmp/podcast";

    private @Mock ItemDownloadManager itemDownloadManager;
    private @Mock PodcastServerParameters podcastServerParameters;
    private @Mock ItemRepository itemRepository;
    private @Mock PodcastBusiness podcastBusiness;
    private @Mock    MimeTypeService mimeTypeService;
    private @InjectMocks ItemBusiness itemBusiness;

    @Before
    public void beforeEach() {
        FileSystemUtils.deleteRecursively(Paths.get(ROOT_FOLDER).toFile());
    }

    @Test
    public void should_find_all_by_page() {
        /* Given */
        PageRequest pageRequest = PageRequest.of(1, 3);
        PageImpl<Item> page = new PageImpl<>(new ArrayList<>());
        when(itemRepository.findAll(any(Pageable.class))).thenReturn(page);

        /* When */
        Page<Item> pageResponse = itemBusiness.findAll(pageRequest);

        /* Then */
        PageAssert.assertThat(pageResponse).isSameAs(page);
        verify(itemRepository, times(1)).findAll(eq(pageRequest));
    }

    @Test
    public void should_save() {
        /* Given */
        Item item = new Item();
        when(itemRepository.save(any(Item.class))).thenReturn(item);
        /* When */
        Item savedItem = itemBusiness.save(item);
        /* Then */
        assertThat(savedItem).isSameAs(item);
        verify(itemRepository, times(1)).save(eq(item));
    }

    @Test
    public void should_find_by_id() {
        /* Given */
        UUID idOfItem = UUID.randomUUID();
        Item item = new Item();
        when(itemRepository.findById(any(UUID.class))).thenReturn(Optional.of(item));
        /* When */
        Item savedItem = itemBusiness.findOne(idOfItem);
        /* Then */
        assertThat(savedItem).isSameAs(item);
        verify(itemRepository, times(1)).findById(idOfItem);
    }

    @Test
    public void should_delete() {
        /* Given */
        UUID idOfItem = UUID.randomUUID();
        Podcast podcast = new Podcast();
        podcast.setItems(new HashSet<>());
        Item item = new Item();
        item.setPodcast(podcast);
        podcast.getItems().add(item);

        when(itemRepository.findById(any(UUID.class))).thenReturn(Optional.of(item));

        /* When */
        itemBusiness.delete(idOfItem);
        /* Then */

        verify(itemRepository, times(1)).findById(idOfItem);
        verify(itemDownloadManager, times(1)).removeItemFromQueueAndDownload(eq(item));
        verify(itemRepository, times(1)).delete(eq(item));
        assertThat(podcast.getItems()).isEmpty();
    }

    @Test
    public void should_reindex() throws InterruptedException {
        /* Given */
        /* When */
        itemBusiness.reindex();
        /* Then */
        verify(itemRepository, times(1)).reindex();
    }

    @Test
    public void should_reset_item() {
        /* Given */
        UUID itemId = UUID.randomUUID();
        Item item = mock(Item.class);
        when(item.reset()).thenReturn(item);
        when(itemRepository.findById(any(UUID.class))).thenReturn(Optional.of(item));
        when(itemDownloadManager.isInDownloadingQueue(any(Item.class))).thenReturn(false);
        when(itemRepository.save(any(Item.class))).thenReturn(item);

        /* When */
        Item resetedItem = itemBusiness.reset(itemId);

        /* Then */
        assertThat(resetedItem).isSameAs(item);
        verify(itemRepository, times(1)).findById(eq(itemId));
        verify(itemDownloadManager, times(1)).isInDownloadingQueue(eq(item));
        verify(item, times(1)).reset();
        verify(itemRepository, times(1)).save(eq(item));
    }

    @Test
    public void should_reset_a_downloading_item() {
        /* Given */
        UUID itemId = UUID.randomUUID();
        Item item = mock(Item.class);
        when(itemRepository.findById(any(UUID.class))).thenReturn(Optional.of(item));
        when(itemDownloadManager.isInDownloadingQueue(any(Item.class))).thenReturn(true);

        /* When */
        Item resetedItem = itemBusiness.reset(itemId);

        /* Then */
        assertThat(resetedItem).isNull();
        verify(itemRepository, times(1)).findById(eq(itemId));
        verify(itemDownloadManager, times(1)).isInDownloadingQueue(eq(item));
    }

    @Test
    public void should_find_page_in_podcast() {
        /* Given */
        UUID idPodcast = UUID.randomUUID();
        PageRequest pageRequest = PageRequest.of(0, 20);
        PageImpl<Item> pageOfItem = new PageImpl<>(new ArrayList<>());
        when(itemRepository.findByPodcast(any(UUID.class), any(PageRequest.class))).thenReturn(pageOfItem);

        /* When */
        Page<Item> pageOfPodcast = itemBusiness.findByPodcast(idPodcast, pageRequest);

        /* Then */
        assertThat(pageOfPodcast.getContent()).isEqualTo(new ArrayList<>());
        verify(itemRepository, times(1)).findByPodcast(eq(idPodcast), eq(pageRequest));
    }

    @Test
    public void should_add_item_by_upload() throws IOException, URISyntaxException {
        /* Given */
        UUID idPodcast = UUID.randomUUID();
        MultipartFile uploadedFile = mock(MultipartFile.class);
        Podcast podcast = new Podcast()
                .setDescription("aDescription");
        Long length = 123456789L;
        ROOT_FOLDER = "/tmp/podcast";
        podcast.setTitle("aPodcast");
        String aMimeType = "audio/type";
        String title = "aPodcast - 2015-09-10 - aTitle.mp3";
        Path ITEM_FILE_PATH = Paths.get(ROOT_FOLDER, podcast.getTitle(), title);
        Files.createDirectories(ITEM_FILE_PATH.getParent());
        Files.createFile(ITEM_FILE_PATH);

        when(podcastBusiness.findOne(eq(idPodcast))).thenReturn(podcast);
        when(uploadedFile.getOriginalFilename()).thenReturn(title);
        when(podcastServerParameters.getRootfolder()).thenReturn(Paths.get(ROOT_FOLDER));
        when(itemRepository.save(any(Item.class))).then(i -> i.getArguments()[0]);
        when(podcastBusiness.save(any(Podcast.class))).then(i -> i.getArguments()[0]);
        when(uploadedFile.getSize()).thenReturn(length);
        when(mimeTypeService.getMimeType(anyString())).thenReturn(aMimeType);
        /* When */
        Item item = itemBusiness.addItemByUpload(idPodcast, uploadedFile);

        /* Then */
        assertThat(item)
                .hasTitle("aTitle")
                .hasPubDate(ZonedDateTime.of(LocalDateTime.of(LocalDate.parse(title.split(" - ")[1], DateTimeFormatter.ofPattern("yyyy-MM-dd")), LocalTime.of(0, 0)), ZoneId.systemDefault()))
                .hasUrl(null)
                .hasLength(length)
                .hasMimeType(aMimeType)
                .hasDescription("aDescription")
                .hasFileName(title)
                .hasPodcast(podcast)
                .hasStatus(Status.FINISH);

        assertThat(podcast.getItems()).contains(item);
        verify(podcastServerParameters, times(1)).getRootfolder();
        verify(mimeTypeService, times(1)).getMimeType(eq("mp3"));
        verify(podcastBusiness, times(1)).findOne(eq(idPodcast));
        verify(podcastBusiness, times(1)).save(eq(podcast));
        verify(itemRepository, times(1)).save(eq(item));
    }

    @Test
    public void should_find_by_tags_and_full_text_without_specific_order() {
        /* Given */
        String term = "Foo";
        Set<Tag> tags = io.vavr.collection.HashSet.of(new Tag().setName("Discovery"), new Tag().setName("Fun"));
        PageRequest pageRequest = PageRequest.of(1, 3, Sort.Direction.fromString("DESC"), "title");
        PageImpl<Item> pageResponse = new PageImpl<>(new ArrayList<>());

        when(itemRepository.fullTextSearch(eq(term))).thenReturn(List.of(UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID()));
        when(itemRepository.findAll(any(Predicate.class), any(PageRequest.class))).thenReturn(pageResponse);

        /* When */
        Page<Item> byTagsAndFullTextTerm = itemBusiness.findByTagsAndFullTextTerm(term, tags, Set(Status.FINISH), pageRequest);

        /* Then */
        PageAssert.assertThat(byTagsAndFullTextTerm).isSameAs(pageResponse);
        verify(itemRepository, times(1)).fullTextSearch(eq(term));
        verify(itemRepository, times(1)).findAll(any(Predicate.class), eq(pageRequest));
    }

    @Test
    public void should_find_by_tags() {
        /* Given */
        Set<Tag> tags = io.vavr.collection.HashSet.of(new Tag().setName("Discovery"), new Tag().setName("Fun"));
        PageRequest pageRequest = PageRequest.of(1, 3, Sort.Direction.fromString("DESC"), "title");
        PageImpl<Item> pageResponse = new PageImpl<>(new ArrayList<>());

        when(itemRepository.findAll(any(Predicate.class), any(PageRequest.class))).thenReturn(pageResponse);

        /* When */
        Page<Item> byTagsAndFullTextTerm = itemBusiness.findByTagsAndFullTextTerm("", tags, Set(Status.FINISH), pageRequest);

        /* Then */
        PageAssert.assertThat(byTagsAndFullTextTerm).isSameAs(pageResponse);
        verify(itemRepository, times(1)).findAll(any(Predicate.class), eq(pageRequest));
    }

    @Test
    public void should_find_by_tags_and_full_text_with_pertinence_order_asc() {
        /* Given */
        String term = "Foo";
        Set<Tag> tags = io.vavr.collection.HashSet.of(new Tag().setName("Discovery"), new Tag().setName("Fun"));
        PageRequest pageRequest = PageRequest.of(1, 3, Sort.Direction.fromString("ASC"), "pertinence");
        List<Item> itemsFrom1To20 = IntStream.range(1, 20)
                .mapToObj(id -> new Item().setId(UUID.randomUUID()))
                .collect(List.collector());

        when(itemRepository.fullTextSearch(eq(term))).thenReturn(itemsFrom1To20.map(Item::getId));
        when(itemRepository.findAll(any(Predicate.class))).thenReturn(itemsFrom1To20);

        /* When */
        Page<Item> pageOfItem = itemBusiness.findByTagsAndFullTextTerm(term, tags, Set(Status.FINISH), pageRequest);

        /* Then */
        assertThat(pageOfItem.getContent()).contains(itemsFrom1To20.get(15),itemsFrom1To20.get(14),itemsFrom1To20.get(13));
        verify(itemRepository, times(1)).fullTextSearch(eq(term));
        verify(itemRepository, times(1)).findAll(any(Predicate.class));
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/business/ItemBusinessTest.kt
package com.github.davinkevin.podcastserver.business

import com.github.davinkevin.podcastserver.entity.Item
import com.github.davinkevin.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.manager.ItemDownloadManager
import com.github.davinkevin.podcastserver.service.MimeTypeService
import com.github.davinkevin.podcastserver.service.properties.PodcastServerParameters
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import lan.dk.podcastserver.repository.ItemRepository
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.util.FileSystemUtils
import java.nio.file.Paths
import java.util.*

/**
 * Created by kevin on 02/08/15 for Podcast Server
 */
@ExtendWith(MockitoExtension::class)
class ItemBusinessTest {

    @Mock lateinit var itemDownloadManager: ItemDownloadManager
    @Mock lateinit var podcastServerParameters: PodcastServerParameters
    @Mock lateinit var itemRepository: ItemRepository
    @Mock lateinit var podcastBusiness: PodcastBusiness
    @Mock lateinit var mimeTypeService: MimeTypeService
    @InjectMocks lateinit var itemBusiness: ItemBusiness

    @BeforeEach
    fun beforeEach() {
        FileSystemUtils.deleteRecursively(Paths.get(ROOT_FOLDER).toFile())
    }

    @Test
    fun `should delete`() {
        /* Given */
        val idOfItem = UUID.randomUUID()
        val p = Podcast().apply { items = mutableSetOf() }
        val item = Item().apply { podcast = p }
        p.items!!.add(item)

        whenever(itemRepository.findById(idOfItem)).thenReturn(Optional.of(item))

        /* When */
        itemBusiness.delete(idOfItem)

        /* Then */
        verify(itemRepository, times(1)).findById(idOfItem)
        verify(itemDownloadManager, times(1)).removeItemFromQueueAndDownload(item)
        verify(itemRepository, times(1)).delete(item)
        assertThat(p.items).isEmpty()
    }

//    @Test
//    @Throws(IOException::class, URISyntaxException::class)
//    fun `should add item by upload`() {
//        /* Given */
//        val idPodcast = UUID.randomUUID()
//        val uploadedFile = mock<MultipartFile>()
//        val podcast = Podcast().apply { description = "aDescription"; title = "aPodcast" }
//        val length = 123456789L
//        val aMimeType = "audio/type"
//        val title = "aPodcast - 2015-09-10 - aTitle.mp3"
//        val itemFilePath = Paths.get(ROOT_FOLDER, podcast.title, title)!!
//        Files.createDirectories(itemFilePath.parent)
//        Files.createFile(itemFilePath)
//
//        whenever(podcastBusiness.findOne(idPodcast)).thenReturn(podcast)
//        whenever(uploadedFile.originalFilename).thenReturn(title)
//        whenever(podcastServerParameters.rootfolder).thenReturn(Paths.get(ROOT_FOLDER))
//        whenever(itemRepository.save(any())).then { it.arguments[0] }
//        whenever(podcastBusiness.save(any())).then { it.arguments[0] }
//        whenever(uploadedFile.size).thenReturn(length)
//        whenever(mimeTypeService.getMimeType(anyString())).thenReturn(aMimeType)
//
//        /* When */
//        val item = itemBusiness.addItemByUpload(idPodcast, uploadedFile)
//
//        /* Then */
//        assertThat(item.title).isEqualTo("aTitle")
//        assertThat(item.pubDate).isEqualTo(ZonedDateTime.of(LocalDateTime.of(LocalDate.parse(title.split(" - ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1], DateTimeFormatter.ofPattern("yyyy-MM-dd")), LocalTime.of(0, 0)), ZoneId.systemDefault()))
//        assertThat(item.url).isEqualTo(null)
//        assertThat(item.length).isEqualTo(length)
//        assertThat(item.mimeType).isEqualTo(aMimeType)
//        assertThat(item.description).isEqualTo("aDescription")
//        assertThat(item.fileName).isEqualTo(title)
//        assertThat(item.podcast).isEqualTo(podcast)
//        assertThat(item.status).isEqualTo(Status.FINISH)
//
//        assertThat(podcast.items).contains(item)
//        verify(podcastServerParameters, times(1)).rootfolder
//        verify(mimeTypeService, times(1)).getMimeType("mp3")
//        verify(podcastBusiness, times(1)).findOne(idPodcast)
//        verify(podcastBusiness, times(1)).save(podcast)
//        verify(itemRepository, times(1)).save(item)
//    }

    companion object {
        private const val ROOT_FOLDER = "/tmp/podcast"
    }
}
